api/client/src/testFixtures/java/org/projectnessie/client/auth/oauth2/InteractiveResourceOwnerEmulator.java

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (C) 2023 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.client.auth.oauth2;
+
+import static java.net.HttpURLConnection.HTTP_MOVED_PERM;
+import static java.net.HttpURLConnection.HTTP_MOVED_TEMP;
+import static java.net.HttpURLConnection.HTTP_OK;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.MoreExecutors;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.OutputStream;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.io.PrintStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+abstract class InteractiveResourceOwnerEmulator implements ResourceOwnerEmulator {
+
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(InteractiveResourceOwnerEmulator.class);
+
+  private static final Pattern LOGIN_FORM_ACTION_PATTERN =
+      Pattern.compile("<form.*action=\"([^\"]+)\".*>");
+
+  private static final AtomicInteger COUNTER = new AtomicInteger(1);
+
+  protected final String username;
+  protected final String password;
+
+  private final ExecutorService executor;
+
+  private final PipedOutputStream pipeOut = new PipedOutputStream();
```

Comment:
Hint: these PipedIn/OutputStreams only "try" once per second. Not sure if that's a "problem".

Reply:
It's unfortunate, but not a problem imho. That's how they implemented those pipes :-(

That said, the "writer side" (the auth or device flow) always calls `flush()` after writing something, which in theory should `notifyAll()` and wake up the reader.

Reply:
Yea - if they `flush()`, it's probably fine. I recall that I ran into these "1 second delays" in other situations in the past.